contracts/account/paymaster/PaymasterERC20Guarantor.sol

```diff
@@ -0,0 +1,80 @@
+    using SafeERC20 for IERC20;
+
+    event UserOperationGuaranteed(bytes32 indexed userOpHash, address indexed user, address indexed guarantor);
+
+    function _validatePaymasterUserOp(
+        PackedUserOperation calldata userOp,
+        bytes32 userOpHash,
+        uint256 maxCost
+    ) internal virtual override returns (bytes memory context, uint256 validationData) {
+        // emit additional `UserOperationGuaranteed` event in case there is a guarantor, default to inherited behavior otherwise.
+        (uint256 validationData_, address guarantor) = _fetchGuarantor(userOp);
+        if (validationData_ == ERC4337Utils.SIG_VALIDATION_SUCCESS && guarantor != address(0)) {
+            emit UserOperationGuaranteed(userOpHash, userOp.sender, guarantor);
+        }
+        return super._validatePaymasterUserOp(userOp, userOpHash, maxCost);
+    }
+
+    function _postOp(
+        PostOpMode /* mode */,
+        bytes calldata context,
+        uint256 actualGasCost,
+        uint256 actualUserOpFeePerGas
+    ) internal virtual override {
+        (
+            bytes32 userOpHash,
+            IERC20 token,
+            uint256 prefundAmount,
+            uint256 tokenPrice,
+            address userOpSender,
+            address prefundPayer
+        ) = _decodeContext(context);
+        uint256 actualAmount = _erc20Cost(actualGasCost, actualUserOpFeePerGas, tokenPrice);
+
+        // If the prefundPayer was the userOpSender, refund the remainder.
+        if (prefundPayer == userOpSender) {
+            token.safeTransfer(userOpSender, prefundAmount - actualAmount);
+        }
+        // Otherwise, attempt to deduct the actualAmount from the userOpSender to this paymaster.
+        else if (token.trySafeTransferFrom(userOpSender, address(this), actualAmount)) {
+            // If successful, pay back the guarantor the prefundAmount.
+            token.safeTransfer(prefundPayer, prefundAmount);
+        } else {
+            // Otherwise, refund the guarantor the prefund remainder.
+            token.safeTransfer(prefundPayer, prefundAmount - actualAmount);
+        }
+
+        emit UserOperationSponsored(userOpHash, userOpSender, actualAmount, tokenPrice);
+    }
+
+    function _prefundPayer(PackedUserOperation calldata userOp) internal view virtual override returns (address) {
+        (uint256 validationData, address guarantor) = _fetchGuarantor(userOp);
+        return
+            (validationData == ERC4337Utils.SIG_VALIDATION_SUCCESS && guarantor != address(0))
+                ? guarantor
+                : userOp.sender;
+    }
+
+    function _postOpCost() internal view virtual override returns (uint256) {
+        return super._postOpCost() + 15_000;
+    }
+
+    /**
+     * @dev Fetches the guarantor address and validation data from the user operation.
+     * Must be implemented in order to correctly enable guarantor functionality.
+     */
+    function _fetchGuarantor(
```

Comment:
I'd remove this function entierly to only rely on `_prefundPayer`.

Reply:
I disagree with this. The `_fetchGuarantor` allows to verify whether the guarantor authorized, otherwise if the guarantor just approved the paymaster to spend on his behalf and is hardcoded on a `_prefunPayer` function then it may be drained by the paymaster users.

I would keep it since it also allows more flexibility in who can guarantee. Otherwise if there's just going to be 1 guarantor I'd leave the Paymaster itself to guarantee operations for the users.